atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/TimestampMapping.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.api.watch;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+
+import org.immutables.value.Value;
+
+import com.google.common.collect.Range;
+import com.palantir.lock.watch.LockWatchVersion;
+import com.palantir.logsafe.Preconditions;
+
+@Value.Immutable
+interface TimestampMapping {
+    Map<Long, LockWatchVersion> timestampMapping();
+
+    @Value.Derived
+    default Range<Long> versionRange() {
+        Optional<Long> firstVersion = timestampMapping()
+                .values()
+                .stream()
+                .map(LockWatchVersion::version)
+                .min(Long::compareTo);
+        Optional<Long> lastVersion = timestampMapping()
+                .values()
+                .stream()
+                .map(LockWatchVersion::version)
+                .max(Long::compareTo);
+
+        Preconditions.checkState(firstVersion.isPresent() && lastVersion.isPresent(),
+                "Cannot compute timestamp mapping for empty map of timestamps");
+
+        return Range.closed(firstVersion.get(), lastVersion.get());
+    }
+
+    @Value.Derived
+    default UUID leader() {
+        Optional<UUID> leader = timestampMapping()
```

Comment:
In any case, we don't want to return multiple leaders, because it will just cause those transactions to conflict further down the line anyway - so at best, this would be a retryable exception.

However, as you note, I don't think we should ever hit this, as a new leader will wipe all previous state, including current timestamp mappings (so we won't be able to get it from the cache - if timestamp versions are not present because of this kind of thing, we retry anyway).